scripts/sync_prod

```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
+#
+# sync_prod - 0.1
+#
+#
+# sync_prod is a utility to sync database dumps from production and and to
+# sync media files (if applicable) for a Django application that are
+# user managed (i.e. uploaded photos for a CMS). Assumes that a proper
+# .my.cnf with credentials exists on target hosts for mysqldump.
+#
+# Dependencies:
+# - jq (system install via package manager for JSON parsing)
+#
+# Usage: sync_prod [-h] [-d] config.json
+# -h    Display this help message
+# -d    Dry-run
+
+printUsage() {
+  echo "Usage: ${0} [-h] [-d] config.json"
+  echo "-h    Display this help message"
+  echo "-d    Dry-run"
+}
+
+# Parse usage option flags, with \? as a backstop against invalid options.
+while getopts ":hd" opt; do
+  case ${opt} in
+    h )
+      printUsage
+      exit 0
+      ;;
+    d )
+      dryrun=true
+      ;;
+    \? )
+      echo "Invalid option: -${OPTARG}" 1>&2
+      printUsage 1>&2
+      exit 1
+      ;;
+  esac
+done;
+# Shift past any options that were consumed
+shift $((OPTIND -1))
+
+# Check if there is still a path to config.json
+if [ -z "${1}" ]
+    then
+      echo "No config file supplied."
+      echo ""
+      printUsage 1>&2
+      exit 1
+fi
+
+# Configurations
+config="$1"
+# Use tmp for temporary dumps
+tmp_file=/tmp/db_tmp.sql
+applications=()
+# Get configured storage dir, or if not, assume PWD
+storageDir=$(jq -r ".storageDir" "${config}")
+if [[ storageDir == "null" ]]
+  then
+    storageDir=${PWD}
+fi
+
+# create a storage dir just in case it doesn't exist
+mkdir -p "${storageDir}"
+
+# This loop avoids issues with spaces in values, although they probably
+# shouldn't be there in the first place
+while IFS= read -r line
+do
+  applications+=( "${line}" )
+done < <( jq -c ".applications | .[]" "${config}" )
+
+for i in "${applications[@]}"
+do
+  host=$(echo ${i} | jq -r ".host")
+  db=$(echo ${i} | jq -r ".database")
+  if [[ $host = "null" ]] || [[ $db = "null" ]]
+    then
+      echo "Missing host or database key in ${i}." 1>&2
+      exit 1
+  fi
+  # On a dry run, just say what it ought to connect to
+  # and also what files would be deleted based on the configured
+  # storageDir
+  if  [[ $dryrun ]]
+    then
+      echo "Would connect to host: $host, database: ${db}"
+      find "${storageDir}" -maxdepth 1 -type f -iname "*-${db}.sql" \
```

Comment:
For all dry run functionality: please define (or maybe run, if appropriate) the command in a shared space in the script so that we can guarantee they stay in sync and the dry run will exactly reflect what would actually be done.

Reply:
So pull out the command into its own function so that its dry-run functionality is isolated with it too? Just to be clear that I'm getting what you want.

Reply:
It doesn't matter to me how you do it exactly - I'm just pointing out that right now, e.g. you have the `find` command duplicated, which means that if you edit one you have to remember to edit the other one to keep them in sync. Seems better to define it once for both cases.